jormungandr/src/blockchain/candidate.rs

```diff
@@ -342,8 +342,41 @@ impl CandidateForest {
                     // Find an existing root or create a new one.
                     let fut = future::poll_fn(move || Ok(inner.poll_lock())).and_then(
                         move |mut forest| {
+                            let root_hash = header.hash();
                             let root_parent_hash = header.block_parent_hash();
-                            let (root_hash, is_new) = forest.add_or_refresh_root(header);
+
+                            // skip if the candidate is already present in the tree
+                            let is_new = if forest.candidate_map.contains_key(&root_hash) {
+                                false
+                            } else {
+                                let parent_candidate =
+                                    forest.candidate_map.get_mut(&root_parent_hash);
+
+                                // insert the new candidate correctly if it has a parent in the
+                                // candidate tree
+                                if let Some(parent_candidate) = parent_candidate {
+                                    chain::pre_verify_link(&header, &parent_candidate.header())?;
+
+                                    if parent_candidate.is_applied() {
+                                        forest.add_or_refresh_root(header);
+                                    } else {
+                                        debug_assert!(!parent_candidate
+                                            .children
+                                            .contains(&root_hash));
+
+                                        parent_candidate.children.push(root_hash);
+                                        forest
+                                            .candidate_map
+                                            .insert(root_hash, Candidate::from_header(header));
+                                    }
+                                } else {
+                                    // otherwise just insert a new root
+                                    forest.add_or_refresh_root(header);
+                                }
```

Comment:
This function is already quite long, this could be a member function of the `forest`